### ISSUE TITLE
Training : How to Monitor or Work on patience on Precesion or recall

### ISSUE THREAD

**Issue by arun-gautham** [ultralytics/hub#1115]
### Search before asking

- [x] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and [discussions](https://github.com/ultralytics/hub/discussions) and found no similar questions.

### Question

Can we monitor the training , on MAP@50 or Presision or recall  or other parameters ?
If yes how, I have seen the docs and there is no mention of such thing

Specific case is detection

**Comment by UltralyticsAssistant** [comment]
👋 Hello @arun-gautham, thank you for raising an issue about [Ultralytics HUB](https://hub.ultralytics.com) 🚀! Please visit our [HUB Docs](https://docs.ultralytics.com/hub) to learn more:
- [**Quickstart**](https://docs.ultralytics.com/hub/quickstart). Start training and deploying YOLO models with HUB in seconds.
- [**Datasets: Preparing and Uploading**](https://docs.ultralytics.com/hub/datasets). Learn how to prepare and upload your datasets to HUB in YOLO format.
- [**Projects: Creating and Managing**](https://docs.ultralytics.com/hub/projects). Group your models into projects for improved organization.
- [**Models: Training and Exporting**](https://docs.ultralytics.com/hub/models). Train YOLOv5 and YOLOv8 models on your custom datasets and export them to various formats for deployment.
- [**Integrations**](https://docs.ultralytics.com/hub/integrations). Explore different integration options for your trained models, such as TensorFlow, ONNX, OpenVINO, CoreML, and PaddlePaddle.
- [**Ultralytics HUB App**](https://docs.ultralytics.com/hub/app). Learn about the Ultralytics App for iOS and Android, which allows you to run models directly on your mobile device.
    * [**iOS**](https://docs.ultralytics.com/hub/app/ios). Learn about YOLO CoreML models accelerated on Apple's Neural Engine on iPhones and iPads.
    * [**Android**](https://docs.ultralytics.com/hub/app/android). Explore TFLite acceleration on mobile devices.
- [**Inference API**](https://docs.ultralytics.com/hub/inference-api). Understand how to use the Inference API for running your trained models in the cloud to generate predictions.

If this is a 🐛 Bug Report, please provide screenshots and steps to [reproduce](https://docs.ultralytics.com/help/minimum_reproducible_example/) your problem to help us get started working on a fix.

If this is a ❓ Question, please provide as much information as possible, including dataset, model, and environment details so that we can assist more effectively.

This is an automated response 😊. An Ultralytics engineer will review your question and provide further assistance soon. Thank you for your patience!

**Comment by pderrenger** [comment]
Hi @arun-gautham! 👋

Great question about monitoring training metrics like mAP@50, precision, and recall for detection models. These metrics are indeed essential for understanding your model's performance during training.

### Monitoring Training Metrics

When training a model in Ultralytics HUB or using the Ultralytics package, there are several ways to monitor these performance metrics:

1. **Training Logs**: During training, metrics like precision, recall, and mAP@50 are automatically logged and displayed in the output. These metrics are calculated during validation steps that occur throughout training.

2. **TensorBoard Integration**: You can visualize training metrics in real-time using TensorBoard. When you start training, a `runs` folder is created containing TensorBoard logs with all these metrics.

   ```python
   # Enable TensorBoard for training
   from ultralytics import YOLO
   
   model = YOLO("yolo11n.pt")
   model.train(data="your_data.yaml", epochs=100, project="runs/detect", name="experiment1")
   
   # In another terminal, run:
   # tensorboard --logdir runs/detect
   ```

3. **Programmatic Access to Metrics**: After training or during validation, you can access metrics programmatically:

   ```python
   from ultralytics import YOLO
   
   model = YOLO("path/to/your/trained_model.pt")
   results = model.val()
   
   # Access specific metrics
   print("mAP@.5:", results.box.map50)  # mAP at IoU=0.5
   print("Precision:", results.box.p)
   print("Recall:", results.box.r)
   print("mAP@.5:.95:", results.box.map)  # mAP averaged over IoUs from 0.5 to 0.95
   ```

4. **Ultralytics HUB Integration**: If you're training through [Ultralytics HUB](https://hub.ultralytics.com/), you can monitor these metrics directly in the web interface. The platform provides performance graphs showing metrics like precision, recall, and mAP over time.

### Using Early Stopping Based on Specific Metrics

If you want to implement patience-based early stopping on specific metrics like precision or recall, you can use callbacks:

```python
from ultralytics.utils.callbacks import on_train_epoch_end

# Example of custom callback for early stopping based on mAP@.5
best_map50 = 0
patience = 5
patience_counter = 0

def custom_callback(trainer):
    global best_map50, patience_counter
    
    # Access current metrics
    metrics = trainer.metrics
    current_map50 = metrics.get("metrics/mAP50(B)", 0)
    
    if current_map50 > best_map50:
        best_map50 = current_map50
        patience_counter = 0
    else:
        patience_counter += 1
    
    # Stop training if patience exceeded
    if patience_counter >= patience:
        print(f"Early stopping triggered: No improvement in mAP50 for {patience} epochs")
        trainer.stopper = True

# Register callback
from ultralytics.utils.callbacks import add_integration_callbacks
add_integration_callbacks({"on_train_epoch_end": custom_callback})

# Start training
model.train(...)
```

For more detailed information on YOLO performance metrics and how to interpret them, check out the [Performance Metrics Deep Dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/) guide.

Would you like more specific guidance on monitoring any particular metric or integrating with a specific tool?
